### PR TITLE
Combine prs and nat grib files.

### PR DESCRIPTION
**Pull request by christinaholtNOAA**
This PR adds options to concatenate grib files when multiple paths and file templates are provided. When combined, the resulting file will also remove all duplicated entries. The script self-cleans those combined files upon exit. 

This work also addresses what should happen when a forecast hour is not found. If a grib input file is not found within the waiting period, all accumulated variables will be removed from the subsequent forecast hours, and all other images should be generated. 

The feature that allows accumulated variables to be plotted a single forecast lead time (an -all_leads flag with -f 6, for example), should now work as expected. If combining files in this mode, all files must be present. A real-time feature to wait on the file has been disabled since this this is not how real-time cases are run.

Should address Issues #94, #111, #85, and #84.

### CODE REVIEW

File: create_graphics.py [hunk 1]
```diff
@@ -108,37 +113,29 @@ def create_zip(png_files, zipf):
...
-        first_fcst = 6 if cla.images[0] == 'global' else 1
-        if fcst_hour <= first_fcst:
-            filenames['01fcst'].append(filename)
-        else:
-            filenames['free_fcst'].append(filename)
+    first_fcst = 6 if cla.images[0] == 'global' else 1
+    if fcst_hour <= first_fcst:
+        filenames['01fcst'].append(filename)
+    else:
+        filenames['free_fcst'].append(filename)
```
Comment (christinaholtNOAA) [inline create_graphics.py:130]: This code is now only handling one forecast hour at a time. The loop should take place in the main loop for handling missing files appropriately.

File: create_graphics.py [hunk 2]
```diff
@@ -530,6 +535,198 @@ def parallel_skewt(cla, fhr, ds, site, workdir):
...
+                msg = f'graphics: {variable} {level}'
+                raise errors.NoGraphicsDefinitionForVariable(msg)
+            accumulate = spec.get('accumulate', False)
+
+            if accumulate:
+                print(f'Will not plot {variable}:{level}')
+                cla.images[1][variable].remove(level)
+                if not cla.images[1][variable]:
+                    del cla.images[1][variable]
+
```
Comment (christinaholtNOAA) [inline create_graphics.py:642]: I feel slightly icky about this implementation where I'm changing input arguments explicitly. I may want to clean that up more in a subsequent PR. I think it works okay for now. Happy to hear feedback.

File: create_graphics.py [hunk 3]
```diff
@@ -530,6 +535,198 @@ def parallel_skewt(cla, fhr, ds, site, workdir):
...
+    # Prepare template with all viable forecast hours -- glob accepts *
+    combined_fn = COMBINED_FN.format(fhr=999).replace('999', '*')
+    combined_fp = os.path.join(cla.output_path, combined_fn)
+
+    combined_files = glob.glob(combined_fp)
+
+    print(f'Removing combined files: ')
+    for file_path in combined_files:
+        print(f'  {file_path}')
+        os.remove(file_path)
```
Comment (christinaholtNOAA) [inline create_graphics.py:657]: Since this needs to be done in a couple of places, using a function.

File: create_graphics.py [hunk 4]
```diff
@@ -530,6 +535,198 @@ def parallel_skewt(cla, fhr, ds, site, workdir):
...
+    for tile in tiles:
+        tile_zip_dir = os.path.join(zip_dir, tile)
+        tile_zip_file = os.path.join(tile_zip_dir, 'files.zip')
+        print(f"checking for {tile_zip_file}")
+        if os.path.isfile(tile_zip_file):
+            os.remove(tile_zip_file)
+            print(f"{tile_zip_file} found and removed")
+        os.makedirs(tile_zip_dir, exist_ok=True)
+        zipfiles[tile] = tile_zip_file
+    return zipfiles
```
Comment (christinaholtNOAA) [inline create_graphics.py:685]: This was just moved from the driver function after complaints from the linter.

File: create_graphics.py [hunk 5]
```diff
@@ -530,6 +535,198 @@ def parallel_skewt(cla, fhr, ds, site, workdir):
...
+    '''
+
+    for tile, zipf in zipfiles.items():
+        png_files = glob.glob(os.path.join(workdir, f'*_{tile}_*{fhr:02d}.png'))
+        zip_proc = Process(group=None,
+                           target=create_zip,
+                           args=(png_files, zipf),
+                           )
+        zip_proc.start()
+        zip_proc.join()
```
Comment (christinaholtNOAA) [inline create_graphics.py:729]: Also moved due to linter complaints.

File: create_graphics.py [hunk 6]
```diff
@@ -567,39 +755,65 @@ def graphics_driver(cla):
...
-            print((('-' * 80)+'\n') * 2)
-            print()
-            print(f'Graphics will be created for input file: {grib_path}')
-            print(f'Output graphics directory: {workdir}')
-            print()
-            print((('-' * 80)+'\n') * 2)
+            print(f'{LOG_BREAK}\n',
+                  f'Graphics will be created for input file: {grib_path}\n',
+                  f'Output graphics directory: {workdir} \n'
+                  f'{LOG_BREAK}')
```
Comment (christinaholtNOAA) [inline create_graphics.py:811]: Using this syntax because linter was complaining about "too many statements"

File: create_graphics.py [hunk 7]
```diff
@@ -32,6 +33,11 @@
 
 AIRPORTS = 'static/Airports_locs.txt'
 
+COMBINED_FN = 'combined_{fhr:03d}.grib2'
+TMP_FN = 'combined_{fhr:03d}.tmp.grib2'
```
Comment (christinaholtNOAA) [inline create_graphics.py:37]: Using parameters here because multiple methods need to know how these temporary files are named.

**Review by christinaholtNOAA — COMMENTED** [review]
It may be able to see the diffs in this one with the "split" view in GitHub.

**Review by cshartsough — APPROVED** [review]
These changes seem straightforward, I didn't see any problems.  One minor check:
line 126 has `first_fcst = 6 if cla.images[0] == 'global' else 1`, 
line 762 has `first_fcst = 6 if cla.images[0] == 'global' else 0`,

I assume that's correct but just making sure.
   
Great job, as always.

**Comment by christinaholtNOAA** [comment]
Great question! 

The first instance (`first_fcst = 6 if cla.images[0] == 'global' else 1`) is using the following logic, and we want to catch 0, 1, and 6 for the relevant cases:
```
if fcst_hour <= first_fcst:
        filenames['01fcst'].append(filename)
```


And then the second instance (`first_fcst = 6 if cla.images[0] == 'global' else 0`) has this logic which should start the range from either 0 or 6 depending on the model:
```
for fhr in range(first_fcst, int(cla.fcst_hour[0]), fcst_inc):
```


I think that's right...

**Review by Brian-Jamison — APPROVED** [review]
Great job!  A fairly heavy lift, thanks for taking it on.  Everything looks good.  One thing to note is the typical schedule of the data.  For a HRRR run, the nat files start coming in about 1:03 hrs after the init time, and come in chunks every 10 minutes.  The prs files start coming in about 53 minutes after the init time, and new forecasts come in about every 2-3 minutes, but some can straggle and come in several minutes later.  So some adjustment may be needed for the wait times when combining prs and nat files.

**Comment by christinaholtNOAA** [comment]
@Brian-Jamison That's a great point! I think we will really need to consider how we want to wait and submit these jobs. The code will wait for a while to make sure the real-time files are available, but maybe we want to delay the start of the runs until nat files start showing up.

**Pull request merged**